quasardb/logger.hpp

```diff
@@ -0,0 +1,223 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ *      notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *    * Neither the name of quasardb nor the names of its contributors may
+ *      be used to endorse or promote products derived from this software
+ *      without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY QUASARDB AND CONTRIBUTORS ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE REGENTS AND CONTRIBUTORS BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#pragma once
+
+#include <qdb/log.h>
+#include <pybind11/pybind11.h>
+#include <mutex>
+#include <atomic>
+
+namespace qdb
+{
+  namespace py = pybind11;
+
+  /**
+   * Utility class that wraps around Python's logging module. Should *not*
+   * be used from Python directly, and as such is not registered as a Python
+   * class.
+   *
+   * Note that every log invocations calls a python function, causes a bunch
+   * of reflections, and is generally as slow as you expect it to be. This cannot
+   * be optimized.
+   *
+   * As such, try to keep this logging out of the critical performance path.
+   */
+class logger
+{
+public:
+
+  /**
+   * Default constructor, invoking any .debug() etc on a default initialized
+   * instantiation of this class leads to undefined behavior, but it's needed
+   * for certain situations.
+   */
+  logger() {};
+
+  /**
+   * Acquire a logger instance for a module. This constructor is a relatively
+   * expensive call, try to minimize invocations by re-using a logger accross
+   * invocations.
+   */
+  logger(const std::string & module_name)
+  {
+
+    // TODO(leon): is there a way we can (safely) cache these calls? they should
+    //             remain the same for every python VM instance... is it safe to cache?
+    //
+    // NOTE(leon): looking at the pybind11 code, it appears they never cache anything.. so
+    //             that might hint that it's complex to get right
+    py::module logging = py::module::import("logging");
```

Comment:
Wouldn't it work to make them, hmmm, function static variables?
```suggestion
    static py::module logging = py::module::import("logging");
```

Reply:
Unfortunately, all these Python objects work with a reference to the 'Python Interpreter' under the hood. I don't think there is a strict guarantee that this reference will always be valid, only in the context of a function call.